### 데이터 불러오기

In [1]:
import pandas as pd 

train = pd.read_csv('./data/train.csv')
test  = pd.read_csv('./data/test.csv')
sub   = pd.read_csv('./data/sample_submission.csv') 

#### 원할한 라벨링을 위해 train데이터와 test데이터를 합친다.

In [2]:
test['revenue'] = 0

all = pd.concat([train, test], ignore_index=True)
print(all.shape)
all.iloc[2999:3001]

(7398, 23)


id                              belongs_to_collection    budget  \
2999  3000                                                NaN  35000000   
3000  3001  [{'id': 34055, 'name': 'Pokémon Collection', '...         0   

                                                 genres  \
2999  [{'id': 53, 'name': 'Thriller'}, {'id': 28, 'n...   
3000  [{'id': 12, 'name': 'Adventure'}, {'id': 16, '...   

                                               homepage    imdb_id  \
2999                   http://www.abductionthefilm.com/  tt1600195   
3000  http://www.pokemon.com/us/movies/movie-pokemon...  tt1226251   

     original_language      original_title  \
2999                en           Abduction   
3000                ja  ディアルガVSパルキアVSダークライ   

                                               overview  popularity    ...     \
2999  A young man sets out to uncover the truth abou...   10.512109    ...      
3000  Ash and friends (this time accompanied by newc...    3.851534    ...      

     release_date runtime                                   spoken_languages  \
2999      9/22/11   106.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3000      7/14/07    90.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   

        status                                            tagline  \
2999  Released          They stole his life. He's taking it back.   
3000  Released  Somewhere Between Time & Space... A Legend Is ...   

                             title  \
2999                     Abduction   
3000  Pokémon: The Rise of Darkrai   

                                               Keywords  \
2999  [{'id': 591, 'name': 'cia'}, {'id': 822, 'name...   
3000  [{'id': 11451, 'name': 'pokémon'}, {'id': 115...   

                                                   cast  \
2999  [{'cast_id': 2, 'character': 'Nathan Harper', ...   
3000  [{'cast_id': 3, 'character': 'Tonio', 'credit_...   

                                                   crew   revenue  
2999  [{'credit_id': '5391990d0e0a260fb5001629', 'de...  82087155  
3000  [{'credit_id': '52fe44e7c3a368484e03d683', 'de...         0  

[2 rows x 23 columns]

### 1) production_companies
#### 전처리를 통해여 제작사의 개수를 구한다.
####    
#### 규칙은 제작사 1개당 ,(컴마)는 하나이다. 이는 이름과 id를 구분하기 위해서 쓰인다.
#### 제작사가 2개일 경우는 제작사들끼리 구분하는 ,(컴마)가 하나 더 들어간다.
#### 즉, 제작사의 개수가 1 - 2 - 3 - 4일 경우 ,(컴마)는 1 - 3 - 5 - 7 씩 늘어난다.(제작사 구분 ,(컴마) 하나와 id구분 ,(컴마) 하나 해서 둘씩 증가)

In [ ]:
# ,(컴마)가 두 개부터는 아래 코드가 통하지만 하나일 경우 -1이 되어서 쓸 수 없고,  for문을 사용해야 한다. 

all['Num_of_pcom'] = all['production_companies'].count(',') - 2 

In [4]:
apc = list(all['production_companies'])
apc[0:10]

["[{'name': 'Paramount Pictures', 'id': 4}, {'name': 'United Artists', 'id': 60}, {'name': 'Metro-Goldwyn-Mayer (MGM)', 'id': 8411}]",
 "[{'name': 'Walt Disney Pictures', 'id': 2}]",
 "[{'name': 'Bold Films', 'id': 2266}, {'name': 'Blumhouse Productions', 'id': 3172}, {'name': 'Right of Way Films', 'id': 32157}]",
 nan,
 nan,
 nan,
 "[{'name': 'Ghost House Pictures', 'id': 768}, {'name': 'North Box Productions', 'id': 22637}]",
 nan,
 "[{'name': 'Walt Disney Pictures', 'id': 2}, {'name': 'Jim Henson Productions', 'id': 2504}, {'name': 'Jim Henson Company, The', 'id': 6254}]",
 "[{'name': 'Castle Rock Entertainment', 'id': 97}]"]

In [9]:
for i in apc[0:5]:
    print(i)
    print("")
    print(i.count(','))

[{'name': 'Paramount Pictures', 'id': 4}, {'name': 'United Artists', 'id': 60}, {'name': 'Metro-Goldwyn-Mayer (MGM)', 'id': 8411}]

5
[{'name': 'Walt Disney Pictures', 'id': 2}]

1
[{'name': 'Bold Films', 'id': 2266}, {'name': 'Blumhouse Productions', 'id': 3172}, {'name': 'Right of Way Films', 'id': 32157}]

5
nan



AttributeError: 'float' object has no attribute 'count'

In [7]:
apc[0]

"[{'name': 'Paramount Pictures', 'id': 4}, {'name': 'United Artists', 'id': 60}, {'name': 'Metro-Goldwyn-Mayer (MGM)', 'id': 8411}]"

In [8]:
apc[0].count(',')

5

In [10]:
apc[4]

nan

In [12]:
all[['production_companies', 'production_countries', 'spoken_languages']].head(10)

production_companies  \
0  [{'name': 'Paramount Pictures', 'id': 4}, {'na...   
1        [{'name': 'Walt Disney Pictures', 'id': 2}]   
2  [{'name': 'Bold Films', 'id': 2266}, {'name': ...   
3                                                NaN   
4                                                NaN   
5                                                NaN   
6  [{'name': 'Ghost House Pictures', 'id': 768}, ...   
7                                                NaN   
8  [{'name': 'Walt Disney Pictures', 'id': 2}, {'...   
9  [{'name': 'Castle Rock Entertainment', 'id': 97}]   

                                production_countries  \
0  [{'iso_3166_1': 'US', 'name': 'United States o...   
1  [{'iso_3166_1': 'US', 'name': 'United States o...   
2  [{'iso_3166_1': 'US', 'name': 'United States o...   
3            [{'iso_3166_1': 'IN', 'name': 'India'}]   
4      [{'iso_3166_1': 'KR', 'name': 'South Korea'}]   
5                                                NaN   
6  [{'iso_3166_1': 'US', 'name': 'United States o...   
7                                                NaN   
8  [{'iso_3166_1': 'US', 'name': 'United States o...   
9  [{'iso_3166_1': 'US', 'name': 'United States o...   

                                    spoken_languages  
0           [{'iso_639_1': 'en', 'name': 'English'}]  
1           [{'iso_639_1': 'en', 'name': 'English'}]  
2           [{'iso_639_1': 'en', 'name': 'English'}]  
3  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  
4           [{'iso_639_1': 'ko', 'name': '한국어/조선말'}]  
5           [{'iso_639_1': 'en', 'name': 'English'}]  
6           [{'iso_639_1': 'en', 'name': 'English'}]  
7  [{'iso_639_1': 'ar', 'name': 'العربية'}, {'iso...  
8           [{'iso_639_1': 'en', 'name': 'English'}]  
9           [{'iso_639_1': 'en', 'name': 'English'}]

### 우선은 결측치부터 처리

In [3]:
print('결측치')
print('production_companies : ' , all['production_companies'].isnull().sum())
print('production_countries : ' , all['production_countries'].isnull().sum())
print('spoken_languages     : ' , all['spoken_languages'].isnull().sum())

결측치
production_companies :  414
production_countries :  157
spoken_languages     :  62


#### 1) production_companies

In [4]:
all['production_companies'][0:5]

0    [{'name': 'Paramount Pictures', 'id': 4}, {'na...
1          [{'name': 'Walt Disney Pictures', 'id': 2}]
2    [{'name': 'Bold Films', 'id': 2266}, {'name': ...
3                                                  NaN
4                                                  NaN
Name: production_companies, dtype: object

#### {'name': 'Bold Films', 'id': 2266} 형태로 결측치를 채워본다.
#### [{'name': 'Unkwon', 'id' : 99999}]

In [14]:
unknown = str([{'name': 'Unkwon', 'id' : 99999}])
all['production_companies'] = all['production_companies'].fillna(unknown)
all['production_companies'][0:5]

0    [{'name': 'Paramount Pictures', 'id': 4}, {'na...
1          [{'name': 'Walt Disney Pictures', 'id': 2}]
2    [{'name': 'Bold Films', 'id': 2266}, {'name': ...
3                    [{'name': 'Unkwon', 'id': 99999}]
4                    [{'name': 'Unkwon', 'id': 99999}]
Name: production_companies, dtype: object

In [15]:
print('production_companies : ' , all['production_companies'].isnull().sum())

production_companies :  0


#### 2) production_countries

In [17]:
all['production_countries'][0:10]

0    [{'iso_3166_1': 'US', 'name': 'United States o...
1    [{'iso_3166_1': 'US', 'name': 'United States o...
2    [{'iso_3166_1': 'US', 'name': 'United States o...
3              [{'iso_3166_1': 'IN', 'name': 'India'}]
4        [{'iso_3166_1': 'KR', 'name': 'South Korea'}]
5                                                  NaN
6    [{'iso_3166_1': 'US', 'name': 'United States o...
7                                                  NaN
8    [{'iso_3166_1': 'US', 'name': 'United States o...
9    [{'iso_3166_1': 'US', 'name': 'United States o...
Name: production_countries, dtype: object

In [19]:
all['production_countries'][0]

"[{'iso_3166_1': 'US', 'name': 'United States of America'}]"

In [20]:
all['production_countries'][3]

"[{'iso_3166_1': 'IN', 'name': 'India'}]"

In [21]:
all['production_countries'][4]

"[{'iso_3166_1': 'KR', 'name': 'South Korea'}]"

In [22]:
all['production_countries'][5]

nan

In [28]:
all['production_countries'][70]

"[{'iso_3166_1': 'DE', 'name': 'Germany'}, {'iso_3166_1': 'US', 'name': 'United States of America'}]"

In [31]:
all['production_countries'][125]

"[{'iso_3166_1': 'AU', 'name': 'Australia'}, {'iso_3166_1': 'CN', 'name': 'China'}, {'iso_3166_1': 'DE', 'name': 'Germany'}]"

In [32]:
unknown = str([{'iso_3166_1': 'UC', 'name' : 'Unknown_Contries'}])
# https://www.x2chi.com/366 에서 나라별 약어에 'UC(Unknown Contries)'가 없는 것을 확인함

all['production_countries'] = all['production_countries'].fillna(unknown)
all['production_countries'][0:10] 

0    [{'iso_3166_1': 'US', 'name': 'United States o...
1    [{'iso_3166_1': 'US', 'name': 'United States o...
2    [{'iso_3166_1': 'US', 'name': 'United States o...
3              [{'iso_3166_1': 'IN', 'name': 'India'}]
4        [{'iso_3166_1': 'KR', 'name': 'South Korea'}]
5    [{'iso_3166_1': 'UC', 'name': 'Unknown_Contrie...
6    [{'iso_3166_1': 'US', 'name': 'United States o...
7    [{'iso_3166_1': 'UC', 'name': 'Unknown_Contrie...
8    [{'iso_3166_1': 'US', 'name': 'United States o...
9    [{'iso_3166_1': 'US', 'name': 'United States o...
Name: production_countries, dtype: object

#### 3) spoken_languages

In [33]:
all['spoken_languages'][0:10]

0             [{'iso_639_1': 'en', 'name': 'English'}]
1             [{'iso_639_1': 'en', 'name': 'English'}]
2             [{'iso_639_1': 'en', 'name': 'English'}]
3    [{'iso_639_1': 'en', 'name': 'English'}, {'iso...
4             [{'iso_639_1': 'ko', 'name': '한국어/조선말'}]
5             [{'iso_639_1': 'en', 'name': 'English'}]
6             [{'iso_639_1': 'en', 'name': 'English'}]
7    [{'iso_639_1': 'ar', 'name': 'العربية'}, {'iso...
8             [{'iso_639_1': 'en', 'name': 'English'}]
9             [{'iso_639_1': 'en', 'name': 'English'}]
Name: spoken_languages, dtype: object

In [42]:
pd.isnull(all['spoken_languages'][1500:1520])  

1500    False
1501    False
1502    False
1503     True
1504    False
1505    False
1506    False
1507    False
1508    False
1509    False
1510    False
1511    False
1512    False
1513    False
1514    False
1515    False
1516    False
1517    False
1518    False
1519    False
Name: spoken_languages, dtype: bool

In [43]:
all['spoken_languages'][1501:1505]

1501             [{'iso_639_1': 'en', 'name': 'English'}]
1502    [{'iso_639_1': 'en', 'name': 'English'}, {'iso...
1503                                                  NaN
1504             [{'iso_639_1': 'en', 'name': 'English'}]
Name: spoken_languages, dtype: object

In [44]:
unknown = str([{'iso_639_1': 'UL', 'name' : 'Unknown_Language'}])
# https://m.blog.naver.com/PostView.nhn?blogId=rahayu&logNo=220691374831&proxyReferer=https%3A%2F%2Fwww.google.com%2F
# 국가별 언어코드 참고, 'UL(Unknown language)'

all['spoken_languages'] = all['spoken_languages'].fillna(unknown)
all['spoken_languages'][1501:1510]

1501             [{'iso_639_1': 'en', 'name': 'English'}]
1502    [{'iso_639_1': 'en', 'name': 'English'}, {'iso...
1503    [{'iso_639_1': 'UL', 'name': 'Unknown_Language'}]
1504             [{'iso_639_1': 'en', 'name': 'English'}]
1505             [{'iso_639_1': 'en', 'name': 'English'}]
1506             [{'iso_639_1': 'en', 'name': 'English'}]
1507    [{'iso_639_1': 'en', 'name': 'English'}, {'iso...
1508             [{'iso_639_1': 'en', 'name': 'English'}]
1509             [{'iso_639_1': 'en', 'name': 'English'}]
Name: spoken_languages, dtype: object

#### 결측치 처리 확인

In [45]:
print('결측치 처리 후')
print('production_companies : ' , all['production_companies'].isnull().sum())
print('production_countries : ' , all['production_countries'].isnull().sum())
print('spoken_languages     : ' , all['spoken_languages'].isnull().sum())

결측치 처리 후
production_companies :  0
production_countries :  0
spoken_languages     :  0


####   
####   
####   
### 다시 production_companies 전처리

#### 규칙은 제작사 1개당 ,(컴마)는 하나이다. 이는 이름과 id를 구분하기 위해서 쓰인다.
#### 제작사가 2개일 경우는 제작사들끼리 구분하는 ,(컴마)가 하나 더 들어간다.
#### 즉, 제작사의 개수가 1 - 2 - 3 - 4일 경우 ,(컴마)는 1 - 3 - 5 - 7 씩 늘어난다.(제작사 구분 ,(컴마) 하나와 id구분 ,(컴마) 하나 해서 둘씩 증가)

In [ ]:
# ,(컴마)가 두 개부터는 아래 코드가 통하지만 하나일 경우 -1이 되어서 쓸 수 없고,  for문을 사용해야 한다. 

all['Num_of_pcom'] = all['production_companies'].count(',') - 2 

In [47]:
apc = list(all['production_companies'])
for i in apc[0:10]:
    print(i)
    print("")
    print(i.count(','))

[{'name': 'Paramount Pictures', 'id': 4}, {'name': 'United Artists', 'id': 60}, {'name': 'Metro-Goldwyn-Mayer (MGM)', 'id': 8411}]

5
[{'name': 'Walt Disney Pictures', 'id': 2}]

1
[{'name': 'Bold Films', 'id': 2266}, {'name': 'Blumhouse Productions', 'id': 3172}, {'name': 'Right of Way Films', 'id': 32157}]

5
[{'name': 'Unkwon', 'id': 99999}]

1
[{'name': 'Unkwon', 'id': 99999}]

1
[{'name': 'Unkwon', 'id': 99999}]

1
[{'name': 'Ghost House Pictures', 'id': 768}, {'name': 'North Box Productions', 'id': 22637}]

3
[{'name': 'Unkwon', 'id': 99999}]

1
[{'name': 'Walt Disney Pictures', 'id': 2}, {'name': 'Jim Henson Productions', 'id': 2504}, {'name': 'Jim Henson Company, The', 'id': 6254}]

6
[{'name': 'Castle Rock Entertainment', 'id': 97}]

1


#### [{'name': 'Walt Disney Pictures', 'id': 2}, {'name': 'Jim Henson Productions', 'id': 2504}, {'name': 'Jim Henson Company, The', 'id': 6254}]
#### 6
#### 제목에 ,(컴마)가 있어서 다른 방법을 생각해야 한다.